packages/metro/src/DeltaBundler/__tests__/DeltaCalculatorContext-test.js

```diff
@@ -0,0 +1,362 @@
+      reset: false,
+    });
+  });
+
+  it('should retry to build the last delta after getting an error', async () => {
+    await deltaCalculator.getDelta({reset: false, shallow: false});
+
+    fileWatcher.emit('change', {eventsQueue: [{filePath: '/ctx?ctx=xxx'}]});
+
+    traverseDependencies.mockReturnValue(Promise.reject(new Error()));
+
+    await expect(
+      deltaCalculator.getDelta({reset: false, shallow: false}),
+    ).rejects.toBeInstanceOf(Error);
+
+    // This second time it should still throw an error.
+    await expect(
+      deltaCalculator.getDelta({reset: false, shallow: false}),
+    ).rejects.toBeInstanceOf(Error);
+  });
+
+  it('should never try to traverse a file after deleting it', async () => {
+    await deltaCalculator.getDelta({reset: false, shallow: false});
+
+    // First modify the file
+    fileWatcher.emit('change', {eventsQueue: [{filePath: '/ctx?ctx=xxx'}]});
+
+    // Then delete that same file
+    fileWatcher.emit('change', {
+      eventsQueue: [{type: 'delete', filePath: '/ctx?ctx=xxx'}],
+    });
+
+    traverseDependencies.mockReturnValue(
+      Promise.resolve({
+        added: new Map(),
+        modified: new Map([['/bundle', entryModule]]),
+        deleted: new Set(['/ctx?ctx=xxx']),
+      }),
+    );
+
+    expect(
+      await deltaCalculator.getDelta({reset: false, shallow: false}),
+    ).toEqual({
+      added: new Map(),
+      modified: new Map([['/bundle', entryModule]]),
+      deleted: new Set(['/ctx?ctx=xxx']),
+      reset: false,
+    });
+
+    expect(traverseDependencies).toHaveBeenCalledTimes(1);
+    expect(traverseDependencies.mock.calls[0][0]).toEqual(['/bundle']);
+  });
+
+  it('does not traverse a file after deleting it and one of its dependencies', async () => {
+    await deltaCalculator.getDelta({reset: false, shallow: false});
+
+    // Delete a file
+    fileWatcher.emit('change', {
+      eventsQueue: [{type: 'delete', filePath: '/ctx?ctx=xxx'}],
+    });
+
+    // Delete a dependency of the deleted file
+    fileWatcher.emit('change', {
+      eventsQueue: [{type: 'delete', filePath: '/foo'}],
+    });
+
+    traverseDependencies.mockReturnValue(
+      Promise.resolve({
+        added: new Map(),
+        modified: new Map([['/bundle', entryModule]]),
+        deleted: new Set(['/ctx?ctx=xxx']),
+      }),
+    );
+
+    await deltaCalculator.getDelta({reset: false, shallow: false});
+
+    // Only the /bundle module should have been traversed (since it's an
+    // inverse dependency of /ctx?ctx=xxx).
+    expect(traverseDependencies).toHaveBeenCalledTimes(1);
+    expect(traverseDependencies.mock.calls[0][0]).toEqual(['/bundle']);
+  });
+
+  it('should not do unnecessary work when adding a context module file after deleting it', async () => {
+    await deltaCalculator.getDelta({reset: false, shallow: false});
+
+    // First delete a file
+    fileWatcher.emit('change', {
+      eventsQueue: [{type: 'delete', filePath: '/ctx?ctx=xxx'}],
```

Comment:
I don't get what this test is trying to simulate: the filesystem watcher will never emit events for `/ctx?ctx=xxx` because there is no such file on disk. I'll try to rewrite this test case to cover a more realistic scenario.